File: flutter-sdk/lib/src/data/repositories/chat_message/chat_message_repository.dart
// Copyright (c) Yalochat, Inc. All rights reserved.

// TODO move this repository to a public folder so it can be extended
// and SDK user could implement their own storage solution
import 'package:chat_flutter_sdk/src/common/page.dart';
import 'package:chat_flutter_sdk/src/common/result.dart';
import 'package:chat_flutter_sdk/src/domain/models/chat_message/chat_message.dart';

// A repository that handles chat messages in a data source.
abstract class ChatMessageRepository {
  // Method to get chat pages in descending order, a null cursor provides the first page
  // subsequent calls should include the cursor returned inside the PageInfo
  Future<Result<Page<ChatMessage>>> getChatMessagePageDesc(
    int? cursor,
    int pageSize,
  );

  // Inserts a chat message to the data source.
  Future<Result<ChatMessage>> insertChatMessage(ChatMessage message);


  // Updates a chat message in storage
  Future<Result<bool>> replaceChatMessage(ChatMessage message);

  // TODO: Add clear messages
}

File: web-sdk/src/data/repositories/chat-message/chat-message-repository.ts
// Copyright (c) Yalochat, Inc. All rights reserved.

import type { Page } from '@domain/common/page';
import type { Result } from '@domain/common/result';
import type { ChatMessage } from '@domain/models/chat-message/chat-message';

export interface ChatMessageRepository {
  // Returns a page of messages in descending order. Pass null cursor for the first page;
  // subsequent calls should use the nextCursor from the returned PageInfo.
  getChatMessagePageDesc(
    cursor: number | null,
    pageSize: number
  ): Promise<Result<Page<ChatMessage>>>;

  // Inserts a message and returns it with the assigned id.
  insertChatMessage(message: ChatMessage): Promise<Result<ChatMessage>>;

  // Replaces an existing message by id.
  replaceChatMessage(message: ChatMessage): Promise<Result<boolean>>;
}
